### PR TITLE
[SPARK-30408][SQL] Should not remove orderBy in sortBy clause in Optimizer

### PR DESCRIPTION
**Pull request by apapi**
### What changes were proposed in this pull request?
Fix defect [SPARK-30408](https://issues.apache.org/jira/browse/SPARK-30408) in EliminateSorts:  orderBy in sortBy clause was removed by EliminateSorts.
code to reproduce:
```
val dataset = Seq( ("a", 1, 4), ("b", 2, 5), ("b2", 2, 2), ("c", 3, 6) ).toDF("a", "b", "c") 
val groupData = dataset.orderBy("b")
val sortData = groupData.sortWithinPartitions("c")
```
The content of groupData is:
```
partition 0: 
    [a,1,4]
partition 1: 
    [b,2,5]
    [b2,2,2]
partition 2: 
    [c,3,6]
```
The content of sortData is:
```
partition 0: 
    [a,1,4]
    [b,2,5]
partition 1: 
    [b2,2,2]
    [c,3,6]
```
The content of sortData is not correct because the repartition operator introduced by orderBy was removed by EliminateSorts.
The content of sortData should be.
```
partition 0: 
    [a,1,4]
partition 1: 
    [b2,2,2]
    [b,2,5]
partition 2: 
    [c,3,6]
```

### Why are the changes needed?
This PR fixed defect [SPARK-30408](https://issues.apache.org/jira/browse/SPARK-30408).
Without this fix, the partition of 
    ```rdd.orderBy("b").sortWithinPartitions("c")```
is same as 
    ```rdd.sortWithinPartitions("c")```
which is not correct.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
I add an UT in ```EliminateSortsSuite``` to test this patch.

### CODE REVIEW

**Comment by maropu** [comment]
Can you use the new PR template? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE
And then, can you describe more in the PR description? e.g., what's the issue, how did you fix it, why do we need this fix, brabrabra...

**Comment by maropu** [comment]
ok to test

**Comment by apapi** [comment]
> Can you use the new PR template? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE
> And then, can you describe more in the PR description? e.g., what's the issue, how did you fix it, why do we need this fix, brabrabra...

**Comment by maropu** [comment]
> The content of sortData should be same as groupData.

I might miss your point, but why do you think so?
It seems the plans below looks ok to me;
```
scala> groupData.explain(true)
== Parsed Logical Plan ==
'Sort ['b ASC NULLS FIRST], true
+- Project [_1#3 AS a#10, _2#4 AS b#11, _3#5 AS c#12]
   +- LocalRelation [_1#3, _2#4, _3#5]

== Analyzed Logical Plan ==
a: string, b: int, c: int
Sort [b#11 ASC NULLS FIRST], true
+- Project [_1#3 AS a#10, _2#4 AS b#11, _3#5 AS c#12]
   +- LocalRelation [_1#3, _2#4, _3#5]

== Optimized Logical Plan ==
Sort [b#11 ASC NULLS FIRST], true
+- LocalRelation [a#10, b#11, c#12]

== Physical Plan ==
*(1) Sort [b#11 ASC NULLS FIRST], true, 0  /// <==== global sort
+- Exchange rangepartitioning(b#11 ASC NULLS FIRST, 200), true, [id=#7]
   +- LocalTableScan [a#10, b#11, c#12]


scala> sortData.explain(true)
== Parsed Logical Plan ==
'Sort ['c ASC NULLS FIRST], false
+- Sort [b#11 ASC NULLS FIRST], true
   +- Project [_1#3 AS a#10, _2#4 AS b#11, _3#5 AS c#12]
      +- LocalRelation [_1#3, _2#4, _3#5]

== Analyzed Logical Plan ==
a: string, b: int, c: int
Sort [c#12 ASC NULLS FIRST], false
+- Sort [b#11 ASC NULLS FIRST], true
   +- Project [_1#3 AS a#10, _2#4 AS b#11, _3#5 AS c#12]
      +- LocalRelation [_1#3, _2#4, _3#5]

== Optimized Logical Plan ==
Sort [c#12 ASC NULLS FIRST], false
+- LocalRelation [a#10, b#11, c#12]

== Physical Plan ==
*(1) Sort [c#12 ASC NULLS FIRST], false, 0 /// <==== local sort
+- *(1) LocalTableScan [a#10, b#11, c#12]
```

**Comment by apapi** [comment]
@maropu 
My example in description is ambiguous, I updated the example.

More details:
For 
```val sortData = dataset.orderBy("b").sortWithinPartitions("c")```

In Analyzed Logical Plan of sortData, there are 2 Sorts, 
the first is is a local sort, which is the result of ```sortWithinPartitions("c")```.
The second is a global sort, which is the result of ```orderBy("b")```.

But in the Optimized Logical plan, there is only one Sort, it is the first local sort, which is the result of ```sortWithinPartitions("c")```. The global sort (which is the result of ```orderBy("b")```) was removed by the Optimizer.

The dataset was first repartitioned by ```orderBy()```, then content within each partition sorted by ```sortWithinPartitions()```.
```orderBy()``` may lead to repartition, but ```sortWithinPartitions()``` will not.
So remove global sort lead to the lost of the repartition operation, which is not correct.

**Comment by maropu** [comment]
Ah, I got the point.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -980,15 +980,25 @@ object EliminateSorts extends Rule[LogicalPlan] {
...
+        recursiveRemoveSort(child)
+      } else {
+        s.copy(child = recursiveRemoveSort(child))
+      }
+    case _ => recursiveRemoveSort(plan)
+  }
+
   private def recursiveRemoveSort(plan: LogicalPlan): LogicalPlan = plan match {
-    case Sort(_, _, child) => recursiveRemoveSort(child)
+    case s @ Sort(_, _, child) => recursiveRemoveSort(s, child)
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: How about the case , `global sort => project/filter => local sort`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -980,15 +980,25 @@ object EliminateSorts extends Rule[LogicalPlan] {
...
+    case s @ Sort(_, _, child) => s.copy(child = recursiveRemoveSort(s, child))
     case j @ Join(originLeft, originRight, _, cond, _) if cond.forall(_.deterministic) =>
       j.copy(left = recursiveRemoveSort(originLeft), right = recursiveRemoveSort(originRight))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
       g.copy(child = recursiveRemoveSort(originChild))
   }
 
+  private def recursiveRemoveSort(parent: Sort, plan: LogicalPlan): LogicalPlan = plan match {
+    case s @ Sort(_, _, child) =>
+      if (parent.global == s.global) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: How about the case , `global sort => project/filter => local sort`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 3]
```diff
@@ -308,4 +308,11 @@ class EliminateSortsSuite extends PlanTest {
     val correctAnswer = PushDownOptimizer.execute(noOrderByPlan.analyze)
     comparePlans(optimized, correctAnswer)
   }
+
+  test("should not remove orderBy in sortBy clause") {
+    val plan = testRelation.orderBy('a.asc).sortBy('b.desc)
+    val optimized = Optimize.execute(plan.analyze)
+    val correctAnswer = testRelation.orderBy('a.asc).sortBy('b.desc).analyze
+    comparePlans(optimized, correctAnswer)
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: Can you add more tests (e.g., more chained cases with filters/projects, `global sort => filters/projects => local sort => local sort => global sort ....`)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (apapi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I add new UT for this.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala (hunk 3 above)
Comment (apapi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: I add new UT for this.

**Comment by apapi** [comment]
@maropu Any thing I need to do for "unknown error code, -9"?

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala [hunk 4]
```diff
@@ -308,4 +308,27 @@ class EliminateSortsSuite extends PlanTest {
     val correctAnswer = PushDownOptimizer.execute(noOrderByPlan.analyze)
     comparePlans(optimized, correctAnswer)
   }
+
+  test("should not remove orderBy in sortBy clause") {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: @apapi . Please change like the following.
```scala
- test("should not remove orderBy in sortBy clause") {
+ test("SPARK-30408 should not remove orderBy in sortBy clause") {
```

**Comment by dongjoon-hyun** [comment]
cc @gatorsmile and @cloud-fan

**Review by dongjoon-hyun — COMMENTED** [review]
In the given example, is the dataset partitioned? For me, it looks like just a plain data frame.
> val dataset = Seq( ("a", 1, 4), ("b", 2, 5), ("b2", 2, 2), ("c", 3, 6) ).toDF("a", "b", "c") 
> val groupData = dataset.orderBy("b")
> val sortData = groupData.sortWithinPartitions("c")

**Comment by prog012** [comment]
@dongjoon-hyun  Yes, the dataset is partitioned as following in the example:
```
partition 0: 
    [a,1,4]
    [b,2,5]
partition 1: 
    [b2,2,2]
    [c,3,6]
```

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala (hunk 4 above)
Comment (apapi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/EliminateSortsSuite.scala]: Done.

**Review by dongjoon-hyun — COMMENTED** [review]
Hmm. I got what you mean. But, if you think that is partitioned, you are saying the following one, right? The user doesn't give us partitioning information yet in your example.
> scala> groupData.rdd.getNumPartitions
> res24: Int = 4

The number of partitioning is 4. It's not 3.

For me, the current implementation seems to follow the following.
> scala> dataset.createOrReplaceTempView("T")
> scala> sql("SELECT * FROM (SELECT * FROM T ORDER BY b) SORT BY c").explain
> == Physical Plan ==
> *(1) Sort [c#12 ASC NULLS FIRST], false, 0
> +- *(1) LocalTableScan [a#10, b#11, c#12]

In this context, the current optimizer removes the inner `ORDER BY`.
`ORDER BY` is supposed to guarantee the total ordering, but `SORT BY` doesn't.
So, the removal looks reasonable in that context.

What you are asking is the following, isn't it?
```
scala> sql("select * from T distribute by b sort by c").explain
== Physical Plan ==
*(1) Sort [c#12 ASC NULLS FIRST], false, 0
+- Exchange hashpartitioning(b#11, 200), false, [id=#111]
   +- LocalTableScan [a#10, b#11, c#12]
```

**Comment by dongjoon-hyun** [comment]
The correct usage for DataSet is the following.
```scala
scala> val groupData = dataset.repartition($"b")
scala> val sortData = groupData.sortWithinPartitions("c")
scala> sortData.explain
== Physical Plan ==
*(1) Sort [c#12 ASC NULLS FIRST], false, 0
+- Exchange hashpartitioning(b#11, 200), false, [id=#16]
   +- LocalTableScan [a#10, b#11, c#12]
```

**Comment by dongjoon-hyun** [comment]
@apapi . I've took a look at this PR for 2.4.5 release, but I'd like to recommend you to close this PR and JIRA (as Invalid). How do you think about that?

cc @maropu , @cloud-fan , @gatorsmile

**Comment by prog012** [comment]
@dongjoon-hyun I can close this PR but I still think this is a defect because it is about optimizer not functionality. The key point here is optimizer should not affect the result of the code.
In the example:
1. The result is incorrect with the optimizer (EliminateSorts), the result is changed because of optimizer even though the code in example is not correct logically.
2. The result of the example is correct if I disable or fix the optimizer in EliminateSorts.

Maybe a better solution is to give hint on the logical error of the example code, but that is beyond this PR.
That's just my two cents.

**Comment by apapi** [comment]
@dongjoon-hyun I can close this PR but I still think this is a defect because it is about optimizer not functionality. The key point here is optimizer should not affect the result of the code.
In the example:
1. The result is incorrect with the optimizer (EliminateSorts), the result is changed because of optimizer even though the code in example is not correct logically.
2. The result of the example is correct if I disable or fix the optimizer in EliminateSorts.

Maybe a better solution is to give hint on the logical error of the example code, but that is beyond this PR.

That's just my two cents.

**Comment by dongjoon-hyun** [comment]
Of course, it does. And, Apache Spark doesn't change the original semantic of `Seq(...).toDF.orderBy("b").sortWithinPartitions`.
> The key point here is optimizer should not affect the result of the code.

As you see, the optimizer doesn't change the result of `Seq(...).toDF.repartition($"b").sortWithinPartitions("c")`, does it?

**Comment by dongjoon-hyun** [comment]
I'll close this PR and JIRA issue. We can continue to discuss and reopen this if needed, @apapi .

**Comment by cloud-fan** [comment]
I think it's about the sort semantic. `df.sort('a)` means the result should be sorted by column `a`. But how about the order of data whose value of column `a` are the same? Can we preserve the original data order? Should `df.sort('b).sort('a)` be different from `df.sort('a)`?

**Pull request closed without merging**